webhooks/virtualmachine/mutation/cdrom_mutator.go

```diff
@@ -0,0 +1,303 @@
+// © Broadcom. All Rights Reserved.
+// The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: Apache-2.0
+
+package mutation
+
+import (
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha5"
+	pkgcfg "github.com/vmware-tanzu/vm-operator/pkg/config"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+	pkgutil "github.com/vmware-tanzu/vm-operator/pkg/util"
+)
+
+// ControllerSlot represents a controller slot position with bus number, unit
+// number, and controller type. It is used to track available slots for CD-ROM
+// devices on IDE and SATA controllers during mutation.
+type ControllerSlot struct {
+	BusNumber      int32
+	UnitNumber     int32
+	ControllerType vmopv1.VirtualControllerType
+}
+
+// MutateCdromControllerOnUpdate mutates CD-ROM controller assignments for
+// VirtualMachine updates. It automatically assigns controller types, bus
+// numbers, and unit numbers for CD-ROM devices when not specified.
+func MutateCdromControllerOnUpdate(
+	ctx *pkgctx.WebhookRequestContext,
+	_ ctrlclient.Client,
+	vm, _ *vmopv1.VirtualMachine) (bool, error) {
+
+	if !pkgcfg.FromContext(ctx).Features.VMSharedDisks {
+		return false, nil
+	}
+
+	if vm.Spec.Hardware == nil {
+		vm.Spec.Hardware = &vmopv1.VirtualMachineHardwareSpec{}
+	}
+
+	var (
+		mutated     bool
+		hwSpec      = vm.Spec.Hardware
+		usedSlotMap = constructUsedSlotMap(vm)
+	)
+
+	for i := range hwSpec.Cdrom {
+		cdrom := &hwSpec.Cdrom[i]
+		var availableSlot *ControllerSlot
+
+		if cdrom.UnitNumber != nil {
+			if cdrom.ControllerBusNumber != nil && cdrom.ControllerType != "" {
+				ctrlID := pkgutil.ControllerID{
+					ControllerType: cdrom.ControllerType,
+					BusNumber:      *cdrom.ControllerBusNumber,
+				}
+				if addNewController(ctrlID, usedSlotMap, hwSpec) {
+					mutated = true
+				}
+			}
+			// Skip if unit number is set but controller information is
```

Comment:
Is this case possible? Do you mean user set unit number while skipping the rest?
I guess we need a story for this case then. Whether we reject them on create or fill out for them

Reply:
There are two cases where this could happen, according to my understanding of the API definition. I listed all the possible combinations here https://github.com/vmware-tanzu/vm-operator/pull/1261/files#r2466472900.

Reply:
Resolved since we will reject these cases in API directly